tests/main_test.go

```diff
@@ -0,0 +1,258 @@
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package tests
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/canonical/pebble/internals/servicelog"
+)
+
+// TestMain builds the pebble binary before running the integration tests.
+func TestMain(m *testing.M) {
+	goBuild := exec.Command("go", "build", "-o", "../pebble", "../cmd/pebble")
+	if err := goBuild.Run(); err != nil {
+		fmt.Println("Cannot build pebble binary:", err)
+		os.Exit(1)
+	}
+
+	exitCode := m.Run()
+	os.Exit(exitCode)
+}
+
+func createLayer(t *testing.T, pebbleDir string, layerFileName string, layerYAML string) {
+	t.Helper()
+
+	layersDir := filepath.Join(pebbleDir, "layers")
+	err := os.MkdirAll(layersDir, 0o755)
+	if err != nil {
+		t.Fatalf("Cannot create layers directory: pipe: %v", err)
+	}
+
+	layerPath := filepath.Join(layersDir, layerFileName)
+	err = os.WriteFile(layerPath, []byte(layerYAML), 0o755)
+	if err != nil {
+		t.Fatalf("Cannot create layers file: %v", err)
+	}
+}
+
+func createIdentitiesFile(t *testing.T, pebbleDir string, identitiesFileName string, identitiesYAML string) {
+	t.Helper()
+
+	identitiesPath := filepath.Join(pebbleDir, identitiesFileName)
+	if err := os.WriteFile(identitiesPath, []byte(identitiesYAML), 0o755); err != nil {
+		t.Fatalf("Cannot create layers file: %v", err)
+	}
+}
+
+func pebbleRun(t *testing.T, pebbleDir string, args ...string) <-chan servicelog.Entry {
+	t.Helper()
+
+	logsCh := make(chan servicelog.Entry)
+
+	cmd := exec.Command("../pebble", append([]string{"run"}, args...)...)
+	cmd.Env = append(os.Environ(), "PEBBLE="+pebbleDir)
+
+	t.Cleanup(func() {
+		err := cmd.Process.Signal(os.Interrupt)
+		if err != nil {
+			t.Errorf("Error sending SIGINT/Ctrl+C to pebble: %v", err)
+		}
+		cmd.Wait()
+	})
+
+	stdoutPipe, err := cmd.StdoutPipe()
+	if err != nil {
+		t.Fatalf("Cannot create stdout pipe: %v", err)
+	}
+	stderrPipe, err := cmd.StderrPipe()
+	if err != nil {
+		t.Fatalf("Cannot create stderr pipe: %v", err)
+	}
+
+	err = cmd.Start()
+	if err != nil {
+		t.Fatalf("Error starting 'pebble run': %v", err)
+	}
+
+	done := make(chan struct{})
+	go func() {
+		defer close(logsCh)
+
+		readLogs := func(parser *servicelog.Parser) {
+			for parser.Next() {
+				if err := parser.Err(); err != nil {
+					t.Errorf("Cannot parse Pebble logs: %v", err)
+				}
+				select {
+				case logsCh <- parser.Entry():
+				case <-done:
+					return
+				}
+			}
+		}
+
+		// Both stderr and stdout are needed, because pebble logs to stderr
+		// while with "--verbose", services otuput to stdout.
+		stderrParser := servicelog.NewParser(stderrPipe, 4*1024)
+		stdoutParser := servicelog.NewParser(stdoutPipe, 4*1024)
+
+		// Channel to signal completion and close logsCh
+		done := make(chan struct{})
```

Comment:
This is confusing: there's a `done` channel defined above, and again here. What are we trying to achieve, and why are there "nested" goroutines here? I'm guessing we want the `readLogs` goroutines to stop when the test finishes? In which case should a `close(done)` go into a `t.Cleanup` func? Perhaps you can clarify or simplify this a bit.